services/wallet/database.go

```diff
@@ -790,31 +789,199 @@ func insertBlocksWithTransactions(creator statementCreator, account common.Addre
 	return nil
 }
 
-func insertRange(creator statementCreator, account common.Address, network uint64, from *big.Int, to *big.Int) (err error) {
-	insert, err := creator.Prepare("INSERT INTO blocks_ranges (network_id, address, blk_from, blk_to) VALUES (?, ?, ?, ?)")
+func upsertRange(creator statementCreator, account common.Address, network uint64, from *big.Int, to *big.Int) (err error) {
+	update, err := creator.Prepare(`UPDATE blocks_ranges
+                SET blk_to = ?
+                WHERE address = ?
+                AND network_id = ?
+                AND blk_to = ?`)
+
 	if err != nil {
 		return err
 	}
 
-	_, err = insert.Exec(network, account, (*SQLBigInt)(from), (*SQLBigInt)(to))
+	res, err := update.Exec((*SQLBigInt)(to), account, network, (*SQLBigInt)(from))
+
+	if err != nil {
+		return err
+	}
+	affected, err := res.RowsAffected()
 	if err != nil {
 		return err
 	}
+	if affected == 0 {
+		insert, err := creator.Prepare("INSERT INTO blocks_ranges (network_id, address, blk_from, blk_to) VALUES (?, ?, ?, ?)")
+		if err != nil {
+			return err
+		}
+
+		_, err = insert.Exec(network, account, (*SQLBigInt)(from), (*SQLBigInt)(to))
+		if err != nil {
+			return err
+		}
+	}
 
 	return
 }
 
+type BlocksRange struct {
+	from *big.Int
+	to   *big.Int
+}
+
+func (db *Database) getOldRanges(account common.Address, network uint64) ([]*BlocksRange, error) {
+	query := `select blk_from, blk_to from blocks_ranges
+	          where address = ?
+	          and network_id = ?
+	          order by blk_from`
+
+	rows, err := db.db.Query(query, account, db.network)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+	ranges := []*BlocksRange{}
+	for rows.Next() {
+		from := &big.Int{}
+		to := &big.Int{}
+		err = rows.Scan((*SQLBigInt)(from), (*SQLBigInt)(to))
+		if err != nil {
+			return nil, err
+		}
+
+		ranges = append(ranges, &BlocksRange{
+			from: from,
+			to:   to,
+		})
+	}
+
+	return ranges, nil
+}
+
+func (db *Database) mergeRanges(account common.Address, network uint64) error {
+	var (
+		tx *sql.Tx
+	)
+
+	ranges, err := db.getOldRanges(account, network)
+	if err != nil {
+		return err
+	}
+
+	log.Info("merge old ranges", "account", account, "network", network, "ranges", len(ranges))
+
+	if len(ranges) <= 1 {
+		return nil
+	}
+
+	tx, err = db.db.Begin()
+	if err != nil {
+		return err
+	}
+
+	defer func() {
```

Comment:
I would use named return values here `(err error)`, because although it works fine in this case, if any code is added and does, something like:
`a, err := ...`
it would shadow `err` and the code below would stop working (for that `err`). 

If you use a named parameter, it would warn you and should be caught by either the compiler or the linter.

Reply:
ok

Reply:
done